cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -0,0 +1,254 @@
+      topicMessageIdStore.persistSubscriberState(topic, lastFetchedMessageId);
+    }
+
+    @Override
+    public void processNotification(DatasetContext context, Notification notification) throws Exception {
+      Map<String, String> properties = notification.getProperties();
+      // Required parameters
+      String programRunIdString = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
+      String programStatusString = properties.get(ProgramOptionConstants.PROGRAM_STATUS);
+
+      // Ignore notifications which specify an invalid ProgramRunId or ProgramRunStatus
+      if (programRunIdString == null || programStatusString == null) {
+        return;
+      }
+
+      ProgramRunStatus runStatus = null;
+      if (programStatusString != null) {
+        try {
+          runStatus = ProgramRunStatus.valueOf(programStatusString);
+        } catch (IllegalArgumentException e) {
+          LOG.warn("Invalid program run status {} passed in notification for program {}",
+                   programStatusString, programRunIdString);
+          return;
+        }
+      }
+
+      final ProgramRunId programRunId = GSON.fromJson(programRunIdString, ProgramRunId.class);
+      final String twillRunId = notification.getProperties().get(ProgramOptionConstants.TWILL_RUN_ID);
+
+      final long stateChangeTimeSecs = getTimeSeconds(notification.getProperties(),
+                                                      ProgramOptionConstants.LOGICAL_START_TIME);
+      final long endTimeSecs = getTimeSeconds(notification.getProperties(), ProgramOptionConstants.END_TIME);
+      final ProgramRunStatus programRunStatus = runStatus;
+      switch(programRunStatus) {
+        case STARTING:
+          String userArgumentsString = properties.get(ProgramOptionConstants.USER_OVERRIDES);
+          String systemArgumentsString = properties.get(ProgramOptionConstants.SYSTEM_OVERRIDES);
+          if (userArgumentsString == null || systemArgumentsString == null) {
+            throw new IllegalArgumentException((userArgumentsString == null) ? "user" : "system" + " arguments was "
+                                               + "not specified in program status notification for program run " +
+                                               programRunId);
+          }
+          if (stateChangeTimeSecs == -1) {
+            throw new IllegalArgumentException("Start time was not specified in program starting notification for " +
+                                                 "program run " + programRunId);
+          }
+          final Map<String, String> userArguments = GSON.fromJson(userArgumentsString, STRING_STRING_MAP);
+          final Map<String, String> systemArguments = GSON.fromJson(systemArgumentsString, STRING_STRING_MAP);
+          Retries.supplyWithRetries(new Supplier<Void>() {
+            @Override
+            public Void get() {
+              store.setStart(programRunId.getParent(), programRunId.getRun(), stateChangeTimeSecs, twillRunId,
```

Comment:
we cannot directly call the store since the store internally also has a transaction, that will cause a nested transaction since `processNotification` is called within a transaction. We have to get the dataset directly.